coral-schema/src/main/java/com/linkedin/coral/schema/avro/SchemaUtilities.java

```diff
@@ -436,48 +439,74 @@ private static boolean isUnionSchemaCompatible(@Nonnull Schema leftSchema, @Nonn
       case INT:
       case LONG:
       case STRING:
-      case NULL:
-        return leftSchema.getType() == rightSchema.getType();
+        return leftSchema.getType() == rightSchema.getType() ? leftSchema : null;
       case FIXED:
         boolean isSameType = leftSchema.getType() == rightSchema.getType();
         boolean isSameNamespace = Objects.equals(leftSchema.getNamespace(), rightSchema.getNamespace());
 
-        return isSameType && (!strictMode || isSameNamespace);
+        return isSameType && (!strictMode || isSameNamespace) ? leftSchema : null;
```

Comment:
If we choose to throw the exception here directly, user cannot know the name of mismatching fields.
By the way, as the name of this method is `getUnionFieldSchema`, I think it is reasonable to use `null` to mean we cannot get the union field schema because of the mismatch.